domain/upgrade/state/state.go

```diff
@@ -78,91 +86,92 @@ func (st *State) SetControllerReady(ctx context.Context, upgradeUUID domainupgra
 		return errors.Trace(err)
 	}
 
-	lookForReadyNodeQuery := `
-SELECT  controller_node_id AS &infoControllerNode.controller_node_id
+	controllerNodeInfo := ControllerNodeInfo{
+		UUID:             uuid.String(),
+		UpgradeInfoUUID:  upgradeUUID.String(),
+		ControllerNodeID: controllerID,
+	}
+
+	checkExistsNodeStmt, err := st.Prepare(`
+SELECT  &ControllerNodeInfo.controller_node_id
 FROM    upgrade_info_controller_node
-WHERE   upgrade_info_uuid = $M.info_uuid
-AND     controller_node_id = $M.controller_id;
-`
-	lookForReadyNodeStmt, err := st.Prepare(lookForReadyNodeQuery, infoControllerNode{}, sqlair.M{})
+WHERE   upgrade_info_uuid = $ControllerNodeInfo.upgrade_info_uuid
+AND     controller_node_id = $ControllerNodeInfo.controller_node_id;
+`, controllerNodeInfo)
 	if err != nil {
-		return errors.Annotatef(err, "preparing %q", lookForReadyNodeQuery)
+		return errors.Annotatef(err, "preparing check exists node statement")
 	}
 
-	insertUpgradeNodeQuery := `
+	insertUpgradeNodeStmt, err := st.Prepare(`
 INSERT INTO upgrade_info_controller_node (uuid, controller_node_id, upgrade_info_uuid)
-VALUES ($M.uuid, $M.controller_id, $M.info_uuid);
-`
-	insertUpgradeNodeStmt, err := st.Prepare(insertUpgradeNodeQuery, sqlair.M{})
+VALUES ($ControllerNodeInfo.*);
+`, controllerNodeInfo)
 	if err != nil {
-		return errors.Annotatef(err, "preparing %q", insertUpgradeNodeQuery)
-	}
-	return errors.Trace(db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
-		err := tx.Query(ctx, lookForReadyNodeStmt, sqlair.M{
-			"info_uuid":     upgradeUUID,
-			"controller_id": controllerID,
-		}).Get(&infoControllerNode{})
-		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+		return errors.Annotatef(err, "preparing insert upgrade node statement")
+	}
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, checkExistsNodeStmt, controllerNodeInfo).Get(&ControllerNodeInfo{})
+		if !errors.Is(err, sqlair.ErrNoRows) {
+			// If the error is not sqlair.ErrNoRows then we have found a row, or
+			// got a different error. If err == nil then the controller node
+			// already exists, so we return nil. Otherwise, we return the error.
```

Comment:
I suspect you're leaning on this comment as a crutch. Could the code be re-written to make this clear?

```
if err == nil {
    return nil
} else if !errors.Is(err, sqlair.ErrNoRows) {
    return errors.Trace(err)
} 

// No row exists, add one (or something)
...
```

Reply:
Or maybe:
```
if err == nil {
    return nil
} else if errors.Id(err, sqlair.ErrNoRows) {
    ...
} else {
   return errors.Trace(err)
}
```

Although this would require an extra level of indentation

Reply:
Changed to the first of these two options and left in a shorter comment explaining why its done like this.